### DISCUSSION TITLE
TM1814 error

### DISCUSSION THREAD

**Discussion by MicePauls** [Makuna/NeoPixelBus#546] [Q&A]
I have a TM1814 strip led, and used this library to program it. But it has this error:

"TM1814_neobus:50:35: error: 'NeoTm1814Method' was not declared in this scope
NeoPixelBus<NeoWrgbTm1814Feature, NeoTm1814Method> strip(PixelCount, PixelPin);"
I'm using Arduio IDE 1.8.13, Could you explain me about it, please?
If you can, you could like give us a simple example for TM1814.
Thank for you support.

**Comment by Makuna** [comment]
What uC are you using?
The TM1814 uses an inverted signal to the normal signal you see for most LEDs.  Some uC have the ability to invert the signal for a pin so the library can do this automatically for you; and thus, there is a `NeoTm1814Method`.  For those uC that can't do this, you need to use `NeoTm1814InvertedMethod` and you then have to provide external inverting chip (or signal transister/fet) to provide the correct signal.

**Reply by MicePauls** [reply]
Thanks you.
My uC is arduino Mega 2560 board,
I tried with and without IC 74hc14 for inverting method, but it didn't effect.
My code was behind, could you show me my fault. Thank you.

#include <NeoPixelBus.h>

const uint16_t PixelCount = 4; // this example assumes 4 pixels, making it smaller will cause a failure
const uint8_t PixelPin = 13;  // make sure to set this to the correct pin, ignored for Esp8266

#define colorSaturation 128

NeoPixelBus<NeoWrgbTm1814Feature, NeoTm1814InvertedMethod> strip(PixelCount, PixelPin);
void setup()
{
    Serial.begin(115200);

    strip.Begin();
    strip.Show();
    delay(100);
    strip.SetPixelSettings(NeoTm1814Settings(65,65,65,380));
    strip.SetPixelColor(0, 0x00000000);
    strip.Show();
}

void loop()
{

}

**Reply by Makuna** [reply]
The AVR series doesn't support inverting, so you will need the external inverting circuit.

When you call SetPixelSettings, you should do with before the first Show() and only really need to do it once.  Further, you are telling what mA current each color element is.  With your code above, you are stating the RGB are 6.5mA and the white is 38mA.  I suspect you have this wrong, but it depends really on the strip you are using.  16mA is a good starting point for most LEDs (so, 160).

Your example only ever sets a black color.   try setting four different colors for the first four pixels.  You may also want to try it the loop
```
strip.SetPixelColor(0, RgbColor(128,0,0)); // red, auto converts to RgbwColor that these strips need
strip.SetPixelColor(1, RgbColor(0,128,0)); // green
strip.SetPixelColor(2, RgbColor(0,0,128)); // blue
strip.SetPixelColor(3, RgbwColor(0,0,0,128)); // white led only using the RgbwColor
strip.Show();
delay(500); // arbitrary delay between color updates
```

**Reply by MicePauls** [reply]
I used a 74HC14 to inverting signal. I checked signal on 74HC14's pin, which was normal. But it was still unaffected. Maybe i'm unlucky with this strip. Thank for your support.
There is my updated code:
#include <NeoPixelBus.h>

const uint16_t PixelCount = 4; // this example assumes 4 pixels, making it smaller will cause a failure
const uint8_t PixelPin = 13;  // make sure to set this to the correct pin, ignored for Esp8266

NeoPixelBus<NeoWrgbTm1814Feature, NeoTm1814InvertedMethod> strip(PixelCount, PixelPin);
void setup()
{
    strip.Begin();
    strip.SetPixelSettings(NeoTm1814Settings(160,160,160,160));
    strip.Show();
    delay(100);

    strip.SetPixelColor(0, RgbColor(255,255,255)); // red, auto converts to RgbwColor that these strips need
    strip.SetPixelColor(1, RgbColor(0,0,0)); // green
    strip.SetPixelColor(2, RgbColor(0,0,0)); // blue
    strip.SetPixelColor(3, RgbwColor(0,0,0,128)); // white led only using the RgbwColor
    strip.Show();
    delay(500);
}

void loop()
{

}

**Reply by Makuna** [reply]
you didn't try moving the setpixels into the loop AND you are only setting the first color.  

NeoPixelBus doesn't continually send out the signal, it only does it on `Show()`; so, if you took 1 second to put your probes on the pin after resetting or power the AVR, then there would not be any signal change.  Further, the TM1814 I believe requires constant signal updates otherwise it powers itself off.  When you cycle the power the AVR, do the same with the strip.

Keep with showing different colors as I mention, the output helps diagnose issues.  What method are you checking the signal on the pin?  Scope?  Volt Meter?  Logic Analyzer? 

And also, make sure to try a different pin, just in case.  Like pin 7 (general purpose open pin on megas).

try this
```
#include <NeoPixelBus.h>

const uint16_t PixelCount = 4; // this example assumes 4 pixels, making it smaller will cause a failure
const uint8_t PixelPin = 13; // make sure to set this to the correct pin, ignored for Esp8266

NeoPixelBus<NeoWrgbTm1814Feature, NeoTm1814InvertedMethod> strip(PixelCount, PixelPin);

void setup() {
  strip.Begin();
  strip.SetPixelSettings(NeoTm1814Settings(160,160,160,160));
  strip.Show();
  delay(50);
}

void loop() {
  strip.SetPixelColor(0, RgbColor(128,0,0)); // red, auto converts to RgbwColor that these strips need
  strip.SetPixelColor(1, RgbColor(0,128,0)); // green
  strip.SetPixelColor(2, RgbColor(0,0,128)); // blue
  strip.SetPixelColor(3, RgbwColor(0,0,0,128)); // white led only using the RgbwColor
  strip.Show();
  delay(50);
}
```

**Reply by MicePauls** [reply]
Thank you so much.
I did it.
For checking invert signal, i used a VOM to measure High, and Low status on 74HC14 IC.
About continue sending data to TM1814, it mean that I can't do other task, other strips, that's right? Or this is a request from Tm1814 IC?
When i used a WS2811 strip led with Fastled library, I need to send data one time, the ICs will keep these data without more sending.

**Reply by Makuna** [reply]
Most "NeoPixel" type LEDs you only need to send once and they HOLD the state.  There are a rare few that require updates periodically or they (the chip) will auto reset.  I believe the TM1814 is one of those. I don't have one to test, so I rely on others experiences and the spec sheet.
You can do other work, you have to split the work up and take less than 500ms (the actual amount is in the spec) before you update again.  You have to code differently than Arduino but look for "cooperative task switching" to read up on the appropriate technique.  I even have a library for it (Task).

**Reply by MicePauls** [reply]
Thank you so much.
